rust/core-lib/src/editor.rs

```diff
@@ -712,6 +717,62 @@ impl Editor {
         */
     }
 
+    fn indent(&mut self) {
+        self.modify_indent(IndentDirection::In);
+    }
+
+    fn outdent(&mut self) {
+        self.modify_indent(IndentDirection::Out);
+    }
+
+    /// Indents lines based on selection, while taking care to preserve cursor position as much as possible.
+    /// Outdent checks if each line in the selection is already fully tabbed (that is, larger than the config's tab size),
+    /// and deletes a tab character at start of line if it is. Otherwise, it checks for the first non-whitespace character and deletes 
+    /// up to that character instead.
+    fn modify_indent(&mut self, direction: IndentDirection) {
+        let mut builder = delta::Builder::new(self.text.len());
+        let tab_text = if self.config.items.translate_tabs_to_spaces {
+                    let tab_size = self.config.items.tab_size;
+                    n_spaces(tab_size)
+            } else {
+                "\t"
+            };
+        for region in self.view.sel_regions() {
+            let (first_line, _) = self.view.offset_to_line_col(&self.text, region.min());
+            let (last_line, last_col) =
+                self.view.offset_to_line_col(&self.text, region.max());
+            let last_line = if last_col == 0 && last_line > first_line {
+                last_line
+            } else {
+                last_line + 1
+            };    
+            for line in first_line..last_line {
```

Comment:
There's a bug here: the same line can have multiple regions, and in this case you will change the indentation once per region, so multiple times for those lines.

I'm not sure what the best approach is. It is guaranteed that regions are in order, so you should be able to keep track of the last line in the last region, and skip that line if it occurs in the next region? Or you could build a set of lines, and then iterate over them.